File: tests/fixtures/ldap_injection/cpp/safe_ldap_search.cpp
// Safe: developer-named sanitize_* helper clears caps on the user value
// before it reaches ldap_search_ext_s.
#include <cstdlib>
#include <ldap.h>

extern const char* sanitize_ldap_filter(const char* raw);

int do_lookup(LDAP* ld) {
    const char* user_filter = std::getenv("USER_FILTER");
    const char* safe = sanitize_ldap_filter(user_filter);
    LDAPMessage* res = nullptr;
    return ldap_search_ext_s(
        ld,
        "ou=people,dc=example,dc=com",
        LDAP_SCOPE_SUBTREE,
        safe,
        nullptr, 0, nullptr, nullptr, nullptr, 0, &res);
}

File: tests/fixtures/ldap_injection/typescript/safe_ldap_search.ts
// Safe: ldap-escape's `filter` helper escapes the user-controlled substring
// before it lands in the filter expression.  Mirrors the unsafe sibling's
// bound-variable shape so only the sanitiser introduction differs.
import * as ldap from 'ldapjs';
import ldapEscape from 'ldap-escape';

const client = ldap.createClient({ url: 'ldap://example.com' });

export function lookup(req: any, res: any) {
    const user = req.query.user;
    const safe = ldapEscape(user);
    const filter = '(uid=' + safe + ')';
    client.search('ou=people,dc=example,dc=com', { filter: filter }, (err: any) => { res.json({ ok: !err }); });
}
